### PR TITLE
[SMTChecker] Fix BMCs constraints on internal functions

### LINKED ISSUE #11885: [SMTChecker] False report on `msg.value` in internal functions
```solidity
contract C {
  function g() external {
    f();
  }
  
  function h() external payable {
    f();
  }

  function f() internal {
    require(msg.value == 0, "");
  }
}
```

However weird this looks, it is valid.

The SMTChecker complains with `BMC: Condition is always true.` on the `msg.value == 0` line.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fix https://github.com/ethereum/solidity/issues/11885

### CODE REVIEW

File: libsolidity/formal/BMC.cpp [hunk 1]
```diff
@@ -188,7 +188,8 @@ bool BMC::visit(FunctionDefinition const& _function)
 	{
 		reset();
 		initFunction(_function);
-		m_context.addAssertion(state().txTypeConstraints() && state().txFunctionConstraints(_function));
+		if (_function.isPublic())
+			m_context.addAssertion(state().txTypeConstraints() && state().txFunctionConstraints(_function));
```
Comment (axic) [inline libsolidity/formal/BMC.cpp:192]: Looks good, but what else is covered by `txTypeConstraints`/`txFunctionConstraints`?

**Comment by leonardoalt** [comment]
Add tests with tx.* and msg.* inside internal functions, also in public functions that are called internally. Note that this is all BMC related, CHC should be correct already.

File: libsolidity/formal/BMC.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp:192]: type constraints on msg.*, tx.*, block.*, msg.value==0 is function is not payable, sets msg.sig and the first 4 bytes of msg.data with the function's signature.

**Comment by leonardoalt** [comment]
Added tests and rebased

**Comment by leonardoalt** [comment]
@axic approve?

File: test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	function g() internal view {
+		assert(data.length >= 0); // should hold
+		assert(uint160(sender) >= 0); // should hold
+		assert(uint32(sig) >= 0); // should hold
+		assert(value >= 0); // should hold
+
+		assert(data.length == msg.data.length); // should hold with CHC
+		assert(sender == msg.sender); // should hold with CHC
+		assert(sig== msg.sig); // should hold with CHC
+		assert(value== msg.value); // should hold with CHC
```
Comment (axic) [inline test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol]: Nit: Weird indentation.

File: test/libsolidity/smtCheckerTests/special/tx_vars_bmc_internal.sol [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+		origin = tx.origin;
+
+		g();
+	}
+	function g() internal view {
+		assert(gas >= 0); // should hold
+		assert(uint160(origin) >= 0); // should hold
+
+		assert(gas == tx.gasprice); // should fail with BMC
+		assert(origin== tx.origin); // should fail with BMC
```
Comment (axic) [inline test/libsolidity/smtCheckerTests/special/tx_vars_bmc_internal.sol]: Here too.

File: test/libsolidity/smtCheckerTests/special/tx_vars_chc_internal.sol [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+		origin = tx.origin;
+
+		g();
+	}
+	function g() internal view {
+		assert(gas >= 0); // should hold
+		assert(uint160(origin) >= 0); // should hold
+
+		assert(gas == tx.gasprice); // should hold with CHC
+		assert(origin== tx.origin); // should hold with CHC
```
Comment (axic) [inline test/libsolidity/smtCheckerTests/special/tx_vars_chc_internal.sol]: And here.

File: test/libsolidity/smtCheckerTests/special/block_vars_bmc_internal.sol [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		assert(dif >= 0); // should hold
+		assert(gas >= 0); // should hold
+		assert(number >= 0); // should hold
+		assert(timestamp >= 0); // should hold
+
+		assert(coin == block.coinbase); // should fail with BMC
+		assert(dif == block.difficulty); // should fail with BMC
+		assert(gas == block.gaslimit); // should fail with BMC
+		assert(number == block.number); // should fail with BMC
+		assert(timestamp== block.timestamp); // should fail with BMC
```
Comment (axic) [inline test/libsolidity/smtCheckerTests/special/block_vars_bmc_internal.sol]: And here.

File: libsolidity/formal/BMC.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/formal/BMC.cpp:192]: So basically here these constraints are not added to internal/external functions?

File: test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol]: Why is it weird? It's just a tab.

Comment (axic) [reply test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol]: `value== msg.value`

File: libsolidity/formal/BMC.cpp (hunk 1 above)
Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp:192]: I think external also goes in here. It shouldn't be added to `internal` functions.

File: test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/special/msg_vars_chc_internal.sol]: Ohh I see. Will fix.

File: libsolidity/formal/BMC.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/formal/BMC.cpp:192]: `isPublic` is true for external?

Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp:192]: `	bool isPublic() const { return visibility() >= Visibility::Public; }`

**Comment by leonardoalt** [comment]
@axic fixed

**Review by axic — APPROVED** [review]

**Pull request merged**